src/backend/utils/resgroup/resgroup.c

```diff
@@ -1925,8 +1925,17 @@ groupAcquireSlot(ResGroupInfo *pGroupInfo, bool isMoveQuery)
 		}
 	}
 
-	/* add into group wait queue */
-	groupWaitQueuePush(group, MyProc);
+	/*
+	 * Add into group wait queue (if not waiting yet).
+	 *
+	 * Need to handle a special case, when MyProc was interrupted
+	 * by SIGTERM while waiting for resource group slot.
+	 * Some callbacks - RemoveTempRelationsCallback for example -
+	 * open new transactions on proc exit. It can cause a double
+	 * add of MyProc to the waiting queue (and its corruption).
+	 */
+	if (!procIsWaiting(MyProc))
```

Comment:
@darthunix what do you think about moving this check inside `groupWaitQueuePush()`? Although at the moment this function is used only in this (current) location, this may change in the future.

If a change to `groupWaitQueuePush()` is made, its contract (loosely described in its comment) must change, too.

Reply:
@leskin-in in fact we already have an equal [assert](https://github.com/arenadata/gpdb/blob/e1357e128c31a6710fad2ecb5e3643fb3dd64b51/src/backend/utils/resgroup/resgroup.c#L3551) inside `groupWaitQueuePush()`, but it takes no effect on production builds without assets. I suppose that anyone using `groupWaitQueuePush()` should inspect its asserts inside at first. After that it becomes obvious, that it is incorrect to handle a waiting proc with this function (or to work [without](https://github.com/arenadata/gpdb/blob/e1357e128c31a6710fad2ecb5e3643fb3dd64b51/src/backend/utils/resgroup/resgroup.c#L3550) exclusive `ResGroupLock)`.

As far as we already have proper asserts inside `groupWaitQueuePush()`, I just fixed its use cases and wrapped with `! procIsWaiting(..)` check. I don't like the idea of replacing asserts with explicit return - such decision can mask some other problems on builds with asserts.